Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -779,6 +779,66 @@ public struct Timestamp: BSONValue, Equatable, Codable {
     }
 }
 
+enum BSONEqualsError: Error { case InvalidArrayArgument(String) }
+
+// See https://github.com/realm/SwiftLint/issues/461
+// swiftlint:disable cyclomatic_complexity
+// A helper function to test equality between two BSONValues. This function tests for exact BSON equality.
+// This means that, for example, differing types with equivalent value are not equivalent.
+// e.g.
+//  4.0 (Double) != 4 (Int)
+func bsonEquals(lhs: BSONValue, rhs: BSONValue) throws -> Bool {
+    validateBSONTypes(lhs: lhs, rhs: rhs)
+
+    switch (lhs, rhs) {
+    case (let l as Int, let r as Int): return l == r
+    case (let l as Int32, let r as Int32): return l == r
+    case (let l as Int64, let r as Int64): return l == r
+    case (let l as Double, let r as Double): return l == r
+    case (let l as Decimal128, let r as Decimal128): return l == r
+    case (let l as Bool, let r as Bool): return l == r
+    case (let l as String, let r as String): return l == r
+    case (let l as RegularExpression, let r as RegularExpression): return l == r
+    case (let l as Timestamp, let r as Timestamp): return l == r
+    case (let l as Date, let r as Date): return l == r
+    case (_ as MinKey, _ as MinKey): return true
+    case (_ as MaxKey, _ as MaxKey): return true
+    case (let l as ObjectId, let r as ObjectId): return l == r
+    case (let l as CodeWithScope, let r as CodeWithScope): return l == r
+    case (let l as Binary, let r as Binary): return l == r
+    case (let l as Document, let r as Document): return l == r
+    case (let l as [BSONValue?], let r as [BSONValue?]):
+        return try zip(l, r).reduce(true, {prev, next in try bsonEquals(lhs: next.0, rhs: next.1) && prev})
```

Comment:
This is a nice functional implementation for the `[BSONValue?]` case, but perhaps making this iterative over the result of `zip(l, r)` is the more clean/readable/Swift-y approach.

I'm not sure what the Swift community's feelings are towards functional implementations like these. It's pretty easily readable for me, but I'm partial to functional code. :)

Reply:
Since `bson_compare` seems like it is calling a function from `mongoc`, it is totally reasonable to imagine that we'd see a performance increase from using it. However, instantiating a new document using the array may be expensive enough to not make a difference. If we can use `bson_compare` directly however, it may be a better move, but it seems like the best we can do currently is use `encode` to append it onto a `Document` and use `Document`'s `==` operator.

In other words, it seems like this boils down to whether the initialization of a `Document` for holding an array for comparison, and its storage, is fast enough to still get gains from calling a `mongoc` function.

We could try doing some low quality benchmarking to see which one wins.

Reply:
Ah, I didn't see @kmahar's response when I posted my response, so there may be some repetition, but seems like this isn't obviously better or worse.

Should we try some simple benchmarking?

Reply:
This would afford you an opportunity to play with the benchmarking facilities in XCTest, so it sounds worthwhile. From a maintainability standpoint, I think the code that wraps the array operands in small documents and uses `==` is more straightforward. Famous last words, but I don't think this will be a particularly hot code path in the driver.

That said, isn't Array already Equatable as of Swift 4.1 (see: [Conditional Conformance](https://swift.org/blog/conditional-conformance/)) if all of its members conform? If so, can we just rely on `==`? Or does this take us back to [SWIFT-226](https://jira.mongodb.org/browse/SWIFT-226), where we can't rely on BSONValue being Equatable since it's not a concrete type?

Reply:
Yes, in 4.1+, arrays in Swift have conditional conformance. However, conditional conformance only works if the members conform to `Equatable` as you've said. The members in this case are `BSONValue`, which _do not_ conform to `Equatable`.

This may make you wonder why we don't just make `BSONValue` conform to this. However, if we did this, then, since arrays are extended to conform to `BSONValue`, doing `arr1 == arr2` would call the implementation for `==` of `BSONValue`, and _not_ trigger conditional conformance. Conditional conformance only kicks in if the container type itself does not conform to `Equatable`.